src/runtime/profiler.cpp

```diff
@@ -429,20 +429,33 @@ __attribute__((destructor))
 WEAK void
 halide_profiler_shutdown() {
     halide_profiler_state *s = halide_profiler_get_state();
-    if (!s->sampling_thread) {
-        return;
+    {
+        // Acquiring the mutex here is essential, as the sampling thread
```

Comment:
The sampling thread only reads current_func, and this thread only writes current_func. I take it TSAN doesn't like this. Non-sampling threads also write current_func via a volatile store, which apparently TSAN is cool with? It's all intentionally racing. Perhaps this just needs to be a call to halide_profiler_set_current_func so that all the stores are at least done in the same way.

Reply:
TSAN assumes that all data races are bad. There are ways to annotate a data race as "tsan should ignore this" but it's hard to do in our runtime environment for various reasons.

FWIW, I found this while trying to track down an intermittent hang when using the `profile` feature with WebAssembly (when `wasm_threads` is also enabled); some cases work just fine but some hang, and so far I haven't found any rhyme or reason to why. TSAN doesn't work under wasm and likely never will (wasm is 32-bit only and TSAN requires 64-bit apparently), but running the hanging case under "native" compilation reported these data races -- sadly, fixing them didn't have any effect on the hang when compiling to wasm, but clean code is good code.

Given that the current design relies on being able to have ignorable data races, I guess I'll just drop this PR.

Reply:
The solution is likely to make the variable atomic. (The issue is if we don't use a mutex or atomic, the reads and writes aren't guaranteed to get barriers, etc.)